README.md
<a href="#" align="">

![](https://i.imgur.com/37MaHi6.png)

</a>

<div align="center">
<p> Tiny package for sending and receiving <code>https(s)</code> requests </p>

[![install size](https://packagephobia.com/badge?p=@splitscript.js/https)](https://packagephobia.com/result?p=@splitscript.js/https)

[![downloads](https://img.shields.io/npm/dm/@splitscript.js/https?color=90ee90&style=flat)](https://www.npmjs.com/package/@splitscript.js/https)

<a href='https://splitscript.js.org/https' style='text-decoration:none;'>

<img src='https://i.imgur.com/8PqPYu0.png' alt='docs' height='100px'>
</a>

</div>

## About

This package is part of [SplitScript.js, the everything framework](https://splitscript.js.org)

It's a tiny package for sending `http(s)` requests

## Install

```bash
$ npm i @splitscript.js/https
```

## Listen for requests

### Start the server

```ts
import https from '@splitscript.js/https'
https.start({ port: 3000 })
```

### Handle requests

Create a file in functions/http/get

```ts
import { Events } from '@splitscript.js/https'
export default async function ({ req, res }: Events.Get) {
	res.write('<html><body><h1>hello world</h1></body></html>')
	res.end()
}
```

### Parse body

Bodys get parsed automatically

```ts
import { Events } from '@splitscript.js/https'
export default async function ({ req, res, bodyParsed }: Events.Post}) {
    res.end(JSON.stringify(bodyParsed))
}
```

### Raw body

To get the raw body

```ts
import { Events } from '@splitscript.js/https'
export default async function ({ req, res, body }: Events.Post}) {
    res.end(body)
}
```

<div align="center">
<sub><code>v2.0.1</code> | <a href='https://splitscript.js.org/https'>docs</a> </sub>

<sub>by [ultraviolet](https://github.com/ultravioletasdf)</sub>

</div>


CHANGELOG.md
# 2.0.1

Fix 404

# 2.0.0

Add a http server

# 1.0.5

Change request response.data to type unknown

# 1.0.0

First version - request, get, post, put, patch, delete

# 1.0.1

`README.md` update

# 1.0.2

Another `README.md` update

# 1.0.3

Fixed type exports in `package.json`

# 1.0.4

params in request options will not stringify if already a string


src/index.ts
import * as https from 'node:https'
import * as http from 'node:http'
import { isHttps, parseData, toQuery } from './utils.js'
import start from './server/start.js'
type Request = http.RequestOptions & {
	body?: any
	params?: object
}

type requestRes = {
	res: http.IncomingMessage
	data: unknown
}
/**
 * Send a `http(s)` request
 *
 * Defaults to `GET` method
 */
export function request(
	url: string,
	options: Request = { method: 'GET' }
): Promise<requestRes> {
	if (!url.startsWith('https://') && !url.startsWith('http://')) {
		url = (options.port === 80 ? 'http://' : 'https://') + url
	}
	const client = isHttps(url, options) ? https : http
	if (options.params) {
		url += toQuery(options.params)
	}
	return new Promise((resolve, reject) => {
		const req = client.request(url, options, (res) => {
			let data = ''

			res.setEncoding('utf-8')
			res.on('data', (chunk: string) => {
				data += chunk
			})

			res.on('end', () => {
				resolve({ res, data: parseData(res, data) })
			})
		})
		req.on('error', reject)
		if (options.body) {
			req.write(Buffer.from(JSON.stringify(options.body)))
		}
		req.end()
	})
}
/** Sends a `GET` request */
export function get(
	url: string,
	options: Request = { method: 'GET' }
): Promise<requestRes> {
	return request(url, { method: 'GET', ...options })
}
/** Sends a `POST` request */
export function post(
	url: string,
	options: Request = { method: 'POST' }
): Promise<requestRes> {
	return request(url, { method: 'POST', ...options })
}
/** Sends a `PUT` request */
export function put(
	url: string,
	options: Request = { method: 'PUT' }
): Promise<requestRes> {
	return request(url, { method: 'PUT', ...options })
}
/** Sends a `PATCH` request */
export function patch(
	url: string,
	options: Request = { method: 'PATCH' }
): Promise<requestRes> {
	return request(url, { method: 'PATCH', ...options })
}
/** Sends a `DELETE` request */
function _delete(
	url: string,
	options: Request = { method: 'DELETE' }
): Promise<requestRes> {
	return request(url, { method: 'DELETE', ...options })
}
export { _delete as delete }
export { start }
export default { request, get, post, put, patch, delete: _delete, start }
export type * as Events from './events'


src/server/start.ts
import http from 'http'
import { URL } from 'node:url'
import { EventEmitter } from '@splitscript.js/core'
import busboy from 'busboy'
import { Readable } from 'stream'
type Options = {
	port?: number
}

export default function start(options: Options = {}) {
	const emitter = new EventEmitter('http', '@splitscript.js/https', [
		'get',
		'post',
		'put',
		'patch',
		'delete',
		'head',
		'options',
		'connect',
		'trace',
		'get/**',
		'post/**',
		'put/**',
		'patch/**',
		'delete/**',
		'head/**',
		'options/**',
		'connect/**',
		'trace/**'
	])
	options.port ??= 3000

	const server = http.createServer((req, res) => {
		let body = ''
		req.on('data', (chunk) => {
			body += chunk.toString()
		})
		req.on('end', async () => {
			let bodyParsed: unknown
			if (req.headers['content-type'] === 'application/json') {
				try {
					bodyParsed = JSON.parse(body)
				} catch {
					bodyParsed = body
				}
			} else if (
				req.headers['content-type'] === 'application/x-www-form-urlencoded'
			) {
				try {
					bodyParsed = new URLSearchParams(body)
				} catch (e) {
					bodyParsed = e
				}
			} else if (req.headers['content-type'] === 'multipart/form-data') {
				try {
					const bb = busboy({ headers: req.headers })
					const formData: Record<string, unknown> = {}
					bb.on('field', (fieldname, val) => {
						formData[fieldname] = val
					})

					// Listen for file events from Busboy
					bb.on(
						'file',
						(
							fieldname: string,
							file: Readable,
							filename: string,
							encoding: string,
							mimetype: string
						): void => {
							const fileChunks: Buffer[] = []

							// Store file data in memory as a Buffer
							file.on('data', (data) => {
								fileChunks.push(data)
							})

							file.on('end', () => {
								const fileData = Buffer.concat(fileChunks)
								formData[fieldname] = {
									filename: filename,
									encoding: encoding,
									mimetype: mimetype,
									data: fileData // Store the file data in memory
								}
							})
						}
					)

					bb.on('finish', () => {
						bodyParsed = formData
					})
				} catch (e) {
					bodyParsed = body
				}
			}
			emitter.send([req.method?.toLowerCase() || 'get'], {
				req,
				res,
				body,
				bodyParsed
			})
			const parsed = new URL(
				req.url ?? '/',
				`http://${req.headers.host ?? '127.0.0.1'}`
			)

			const path = parsed.pathname.replace(/^\/+|\/+$/g, '').split('/')
			emitter.send([req.method?.toLowerCase() || 'get', ...path], {
				req,
				res,
				body,
				bodyParsed
			})
			emitter.listeners().then((listeners) => {
				if (
					!listeners.find((v) => {
						if (v.event.length === 1) {
							return [
								'get',
								'post',
								'put',
								'patch',
								'delete',
								'head',
								'options',
								'connect',
								'trace'
							].includes(req.method?.toLowerCase() || 'get')
						} else {
							return (
								JSON.stringify([
									req.method?.toLowerCase() || 'get',
									...path
								]) === JSON.stringify(v)
							)
						}
					})
				) {
					res.statusCode = 404
					res.end()
				}
			})
		})
	})
	server.listen(options.port, () => {
		console.log(`online on ${options.port}`)
	})
}
